features/step_definitions/table_steps.rb

```diff
@@ -22,6 +22,9 @@
 
 # TODO: can we use "has_table?" instead of this complex step?
 Then /^(?:I |they )?should see (?:the )?following table( with exact columns)?:?$/ do |exact_columns, expected|
+  # wait for the table with the expected data to load
+  find("td", text: expected.rows.first.first)
```

Comment:
Finally figured out why the cucumber didn't work.

So, `wait_for_request` waits for the AJAX calls that have already been triggered.
But in this scenario, due to how React works this wasn't helping, because by the end of the scenario the request was not even sent out.

This is because of the async way of doing things in React - clicking the button didn't immediately send the request, instead it updated the `query` state variable, and on the next cycle of re-rendering by React the `useEffect` hook noticed the change in state and only then made the API call.

Adding a `find` here uses a built-in Capybara waiting mechanism that waits for all the operations to complete and find the element in the DOM.

Reply:
This info is good to go somewhere in a visible place.